components/calendar/src/islamic.rs

```diff
@@ -308,6 +308,276 @@ impl DateTime<IslamicObservational> {
+    fn saudi_criterion(date: RataDie) -> Option<bool> {
+        let sunset = Astronomical::sunset((date - 1).as_moment(), MECCA)?;
+        let tee = Location::universal_from_standard(sunset, MECCA);
+        let phase = Astronomical::lunar_phase(tee);
+        let moonlag = Astronomical::moonlag((date - 1).as_moment(), MECCA)?;
+
+        Some(phase > 0.0 && phase < 90.0 && moonlag > 0.0)
+    }
+
+    pub(crate) fn adjusted_saudi_criterion(date: RataDie) -> bool {
+        if let Some(x) = Self::saudi_criterion(date) {
+            x
+        } else {
+            false
+        }
+    }
+
+    // Closest fixed date on or before date when Saudi visibility criterion is held.
+    // Lisp code reference: https://github.com/EdReingold/calendar-code2/blob/main/calendar.l#L6966
+    fn saudi_new_month_on_or_before(date: RataDie) -> RataDie {
+        let last_new_moon =
+            libm::floor((Astronomical::lunar_phase_at_or_before(0.0, date.as_moment())).inner()); // Gets the R.D Date of the prior new moon
+        let age = date.to_f64_date() - last_new_moon;
+        // Explanation of why the value 3.0 is chosen: https://github.com/unicode-org/icu4x/pull/3673/files#r1267460916
+        let tau = if age <= 3.0 && !Self::adjusted_saudi_criterion(date) {
+            // Checks if the criterion is not yet visibile on the evening of date
+            last_new_moon - 30.0 // Goes back a month
+        } else {
+            last_new_moon
+        };
+
+        next(RataDie::new(tau as i64), Self::adjusted_saudi_criterion) // Loop that increments the day and checks if the criterion is now visibile
+    }
+
+    // Lisp code reference: https://github.com/EdReingold/calendar-code2/blob/main/calendar.l#L6996
+    #[allow(clippy::unwrap_used)]
+    fn saudi_islamic_from_fixed(date: RataDie) -> Date<UmmAlQura> {
+        let crescent = Self::saudi_new_month_on_or_before(date);
+        let elapsed_months =
+            libm::round((crescent - FIXED_ISLAMIC_EPOCH_FRIDAY) as f64 / MEAN_SYNODIC_MONTH);
+        let year = (1.0 + div_rem_euclid_f64(elapsed_months, 12.0).0) as i32;
+        let month = (1.0 + div_rem_euclid_f64(elapsed_months, 12.0).1) as u8;
+        let day = ((date - crescent) + 1) as u8;
+
+        Date::try_new_ummalqura_date(year, month, day).unwrap()
+    }
+
+    // "Fixed" is a day count representation of calendars staring from Jan 1st of year 1 of the Georgian Calendar.
+    // The fixed date algorithms are from
+    // Dershowitz, Nachum, and Edward M. Reingold. _Calendrical calculations_. Cambridge University Press, 2008.
+    //
+    // Lisp code reference: https://github.com/EdReingold/calendar-code2/blob/main/calendar.l#L6981
+    fn fixed_from_saudi_islamic(date: UmmAlQuraDateInner) -> RataDie {
+        let year = date.0.year;
+        let month = date.0.month;
+        let day = date.0.day;
+
+        let midmonth = RataDie::new(libm::floor(
```

Comment:
The Lisp code goes

```lisp
(+ islamic-epoch
  (floor ...
```

The RHS of the assignment statement here is inverting the order of those operations.

Reply:
done !